Mixed PSK 1: Extend PSK-to-MS algorithm in PSA

mprse opened the pull request:
## Description

Extend PSK-to-MS algorithm in PSA to support mixed psk.

Resolves https://github.com/Mbed-TLS/mbedtls/issues/5662

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO  

## Migrations
NO

## Additional comments

Test vectors were generated using python tls library:
https://github.com/python-tls/tls

Scrip can be found below:
[derive_ms.zip](https://github.com/Mbed-TLS/mbedtls/files/8397806/derive_ms.zip)

## Todos
- [x] Implementation
- [x] Tests
- [x] Documentation

superna9999 commented on library/psa_crypto.c:
```diff
@@ -5306,12 +5306,23 @@ static psa_status_t psa_tls12_prf_psk_to_ms_set_key(
      * uint16 with the value N, and the PSK itself.
      */
```
You should update the comment with the forming of the premaster with salt as other_secret in DHE_PSK & RSA_PSK as described in section 3 & 4 of RFC 4279

mprse replied:
Yes, documentation is still TBD.

mprse replied:
Fixed

mpg commented on include/psa/crypto_sizes.h:
```diff
@@ -239,6 +239,25 @@
 
+/** This macro returns the maximum supported length of the other secret for
+ * the TLS-1.2 PSK-to-MS key derivation
+ * (#PSA_ALG_TLS12_PSK_TO_MS(\c hash_alg)).
```
Perhaps also mention that it's the input called "Salt" there?

mpg commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -5185,6 +5185,55 @@ PSA key derivation: TLS 1.2 PSK-to-MS, SHA-384, 0+48
+# sl = salt length in bytes
+PSA key derivation: TLS 1.2 Mix-PSK-to-MS, SHA-256, 48+0, sl 48
+depends_on:PSA_WANT_ALG_SHA_256:PSA_WANT_ALG_TLS12_PSK_TO_MS
+derive_output:PSA_ALG_TLS12_PSK_TO_MS(PSA_ALG_SHA_256):PSA_KEY_DERIVATION_INPUT_SEED:"5bc0b19b4a8b24b07afe7ec65c471e94a7d518fcef06c3574315255c52afe21b5bc0b19b872b9b26508458f03603744d575f463a11ae7f1b090c012606fd3e9f":PSA_KEY_DERIVATION_INPUT_SALT:"ce2fa604b6a3e08fc42eda74ab647adace1168b199ed178dbaae12521d68271d7df56eb56c55878034cf01bd887ba4d7":PSA_KEY_DERIVATION_INPUT_SECRET:"01020304":PSA_KEY_DER ...
```
Where does the data come from? I assume for logs of a successful RSA-PSK handshake (and DHE-PSK) below? It would be good to have a comment about it (or at least mention it in the commit message). Also, we probably should have some from data from an ECDHE-PSK exchange (which gives a smaller salt with the default curve).

mprse replied:
In the PR description I added information how test vectors were generated.

gilles-peskine-arm replied:
Committing the script may be more work than we care to do. This requires making it pass `check-python-files`, which requires adding its dependencies to `ci.requirements.txt`.

I also don't necessarily prefer having a comment inside the file to indicate how the data was generated. It can be difficult to convey precisely which test cases the comment applies to, and there's a fairly high risk that new data will be added or that the data will be modified, causing the scope described in the comment to become wrong.

So my preference is to have instructions for generating the test data in the commit message.

mprse replied:
Sure I can commit this script and provide details in the commit. Oh and still need to add change log entry.

mpg reviewed (changes requested):
Looks pretty good to me, just two minor documentation points.

mprse commented:
Provided the following changes:
- rebased on master (hopefully CI will be happy),
- fixed description of `PSA_TLS12_PSK_TO_MS_OTHER_SECRET_MAX_SIZE` (https://github.com/Mbed-TLS/mbedtls/pull/5685/commits/eb25423c2efad8f11798cefe17be32efd05a9c62)
- Added more test cases and information how to generate test vectors in commit message (https://github.com/Mbed-TLS/mbedtls/pull/5685/commits/6354e818f4f6b17548e10e64bf88e63f76d5fa2a)

superna9999 commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -6900,16 +6900,18 @@ void derive_output( int alg_arg,
                                     MBEDTLS_SVC_KEY_ID_INIT,
-                                    MBEDTLS_SVC_KEY_ID_INIT };
+                                    MBEDTLS_SVC_KEY_ID_INIT,
+                                    MBEDTLS_SVC_KEY_ID_INIT};
```
Missing space before `}` to respect previous indentation

superna9999 reviewed (approved):
Ok for me, only a single feedback about an indentation typo

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.data:
```diff
@@ -5071,127 +5071,201 @@ derive_actions_without_setup:
 
```
I have not reviewed the tests. Please make sure at least the following cases are tested for `PSA_ALG_TLS12_PSK_TO_MS`. (I have not attempted to be exhaustive.)

Good cases:
* [ ] Salt omitted.
* [ ] Salt passed with `input_bytes`, of reasonable length(s).
* [ ] Salt passed with `input_key`, of type `DERIVE`, of reasonable length(s).
* [ ] Salt passed with `key_derivation_key_agreement`.
* [ ] Empty salt passed with `input_bytes`.

Bad cases:
* [ ] Salt passed before seed.
* [ ] Salt passed after secret.
* [ ] Salt passed after label.
* [ ] Salt passed twice.
* [ ] Unreasonably large salt passed with `input_bytes`.
* [ ] Unreasonably large salt passed with `input_key`.
* [ ] Salt of type `RAW_DATA` passed with `input_key`.
* [ ] Pass a secret with `input_bytes` and a salt with `input_key`. `output_key` must fail with `NOT_PERMITTED`.

mprse replied:
I'm working on the new PR where we got additional input (other key) instead of salt for Mixed-PSK.
I stuck with implementation and tests for Mixed-PSK support `key_derivation_key_agreement`. 

According to the documentation:
```
For a DHE-PSKor ECDHE-PSK cipher suite the other secret should be the output of the `PSA_ALG_FFDH` or
`PSA_ALG_ECDH` key agreement performed with the peer. The recommended way to pass this input is to use a key 
derivation algorithm constructed as `PSA_ALG_KEY_AGREEMENT(ka_alg, PSA_ALG_TLS12_PSK_TO_MS(hash_alg))` and to 
call `psa_key_derivation_key_agreement()`.
```

I don't understand how this should work. 
`key_derivation_key_agreement` takes `private key` and `peer key` and produces shared secret that in case of mixed-psk should be the other secret, but how PSK should be provided ?

mpg replied:
I think the documentation is describing the the best way to do this, as it protects intermediate secrets, but that's a bit beyond our immediate goals for TLS. Right now, as an intermediate step, I think it's OK for TLS to keep using `psa_raw_key_agreement()` for the ECDH part, writing the result in the PMS buffer, and then later on get the input from that buffer when deriving the MS. Doing it "the right way" as recommended in the documentation may require non-trivial work such as changing the time when some computations are made, and is out of scope as far as #5663 is concerned, if that's what you were asking about.

Note: you can find an example of usage of `key_derivation_key_agreement` in `test_suite_psa_crypto.function`. (I've never used it myself, so I don't know the answer to you question, but hopefully the test code can help.)

gilles-peskine-arm replied:
> how PSK should be provided

`psa_key_derivation_input_key` or `psa_key_derivation_input_bytes` with `PSA_KEY_DERIVATION_INPUT_SECRET`.

And the (EC)DHE shared secret is passed with `psa_key_derivation_key_agreement` or `psa_raw_key_agreement; psa_key_derivation_input_bytes` with `PSA_KEY_DERIVATION_OTHER_SECRET`.

mprse replied:
Thanks @mpg @gilles-peskine-arm.

So the full flow for mixed-psk with key agreement is:
- input seed,
- key agreement (input other key)
- input psk (secret)
- input label

mprse replied:
Got a question regarding last last bullet in bad cases:
> Pass a secret with input_bytes and a salt with input_key. output_key must fail with NOT_PERMITTED.
If the SALT input comes from a key but the SECRET input comes from input_bytes, output_key will not be permitted, since output_key only cares about where the SECRET input comes from (according to the specification). This is the right choice, so there is no need to change anything in the specification. Please make sure there's a unit test for it.

With my current implementation in this case (secret comes with from input bytes and other secret comes from input key)  key derivation is successful and test passes. I understand that in such scenario `psa_key_derivation_output_bytes()` should return `NOT_PERMITTED` ? Why this operation is not permitted?

gilles-peskine-arm replied:
Schematically, a key derivation combines a secret input with public inputs and produces a potentially secret output. If you use `input_key` with a non-exportable key for the secret input and `output_key` for the secret output, the secret output cannot be reconstructed outside the crypto service. On the other hand, if you use `input_bytes` for the secret input, the output can be reconstructed outside the crypto service, by making the same calculation from the inputs. `output_key` is reserved for the case where the secret input comes from `input_key` to reduce the risk of doing a security review and thinking “my key comes from `output_key` so I know it can't be reconstructed outside the crypto service”, at least on systems where all keys are non-exportable.

gilles-peskine-arm reviewed (changes requested):
I have only done an architecture review (I only looked at the headers).

I'm ok with the design, but I have some reservations about the choice of input step constant, and some comments about the documentation.

I've made a corresponding pull request to the PSA specification ([private link](https://github.com/ARM-software/psa-crypto-api/pull/549)). There I currently use a new input step constant, as discussed [here](), but I'm open to changing this.

mpg commented on include/psa/crypto_sizes.h:
```diff
@@ -239,6 +239,27 @@
+ *
+ * Define 384 bytes space for the other secret.
+ */
+#define PSA_TLS12_PSK_TO_MS_OTHER_SECRET_MAX_SIZE 384
```
I agree with the goal (use a smaller max size if we can), but not really with the proposal. Two things:

1. For RSA, I don't think we care about the key size - the size of the `other_secret` part (that comes from RSA decryption) is fixed to 48 bytes by the RFC, so we should use 48 here.
2. For FFDH, as you said, the FFDH key agreement is currently not done by PSA, so I'm worried about configurations where `MBEDTLS_DHM_C` would be enabled but not `PSA_WANT_ALG_FFDH` - wouldn't that result in `PSA_RAW_KEY_AGREEMENT_OUTPUT_MAX_SIZE` being too small?

mprse replied:
Maybe we don't need this macro and instead we can allocate the memory for ``pms` dynamically and free it at the end of the `psa_tls12_prf_psk_to_ms_set_key` function.

mpg commented on library/psa_crypto.c:
```diff
@@ -5197,11 +5204,34 @@ static psa_status_t psa_tls12_prf_set_seed( psa_tls12_prf_key_derivation_t *prf,
+
+    if( data_length != 0 )
+    {
+        prf->salt = mbedtls_calloc( 1, data_length );
```
On second though, why do we use a dynamic buffer here, why we use a statically-size one for the PMS, which is larger? This one is part of the `psa_tls12_prf_key_derivation_t` while the PMS buffer is local to a function, but does it justify the difference? (This is a question, not necessarily asking for a change.)

mprse replied:
It's done in this way because of consistency with the other inputs (seed, secret, etc.). For PMS static buffer was used. 

From my perspective if there is a problem with the optimal size of the static buffer for `PMS` it should be allocated dynamically.

mpg replied:
> From my perspective if there is a problem with the optimal size of the static buffer for PMS it should be allocated dynamically.

Yes, I tend to agree. Since we're already doing dynamic allocation as part of the operation anyway, let's keep advantage of it and avoid the thorny sizing issues associated with a static buffer.

mpg commented on library/psa_crypto.c:
```diff
@@ -5268,25 +5300,51 @@ static psa_status_t psa_tls12_prf_psk_to_ms_set_key(
+        *cur++ = MBEDTLS_BYTE_1( prf->salt_length );
+        *cur++ = MBEDTLS_BYTE_0( prf->salt_length );
+        memcpy( cur, prf->salt, prf->salt_length );
+        cur += prf->salt_length;
```
Actually I think we could clear & free `salt` now, (or before exiting the function, at the same time as we clear the PMS). Doesn't make much of a difference, but as a general rule it's good practice to clear secrets as soon as possible.

mprse replied:
I agree, but on the other hand it is good free all buffers in the same place (probably other inputs could also be freed earlier).

mpg commented on include/psa/crypto_values.h:
```diff
@@ -1814,10 +1814,21 @@
  * passed in the order given here:
  * - #PSA_KEY_DERIVATION_INPUT_SEED is the seed.
- * - #PSA_KEY_DERIVATION_INPUT_SECRET is the secret key.
+ * - #PSA_KEY_DERIVATION_INPUT_SALT is the the output of RSA
```
I felt conflicted about this too. Thanks for the point about key type, I had failed to consider that. I think this makes me more inclined to use a new, dedicated input type, then, if input type proliferation is not considered an issue.

mprse replied:
Reading all the comments about how other secret input should be provided (using existing `SALT` input vs new `OTHER_SECRET` input) I'm getting confused.

> On the gripping hand, using SALT requires a change to the specification of PSA_KEY_DERIVATION_INPUT_SALT, because when using input_key, that requires an input of type RAW_DATA or PEPPER. But for TLS12_PSK_TO_MS, the input type should be DERIVE instead. (Please ensure there are tests for this.) Arguably, SALT might permit DERIVE inputs for any KDF, but I fear confusion between inputs for other KDF then.

https://github.com/Mbed-TLS/mbedtls/pull/5685/files#r844122587

Currently the PSK is provided using `psa_key_derivation_input_key()` where where key needs to be of DERIVE type (nothing changes here).
Other key is provided as salt using `psa_key_derivation_input_bytes()`, so other key is provided as raw data.
I don't see the issue here.

mpg replied:
@mprse I'm not sure if you have access to the [PSA PR](https://github.com/ARM-software/psa-crypto-api/pull/549) but in case you don't: the discussion there seems to have converged to using a new type called `PSA_KEY_DERIVATION_INPUT_OTHER_SECRET`.

gilles-peskine-arm replied:
> Currently the PSK is provided using psa_key_derivation_input_key() where where key needs to be of DERIVE type (nothing changes here).
Other key is provided as salt using psa_key_derivation_input_bytes(), so other key is provided as raw data.

All inputs can be provided either with `input_key` or with `input_bytes`, at the application's choice. With `input_key`, there are [constraints of type](https://armmbed.github.io/mbed-crypto/html/api/ops/kdf.html?highlight=type#c.PSA_KEY_DERIVATION_INPUT_SALT). Some inputs can be provided with `key_agreement` (https://armmbed.github.io/mbed-crypto/html/api/ops/kdf.html?highlight=psa_key_derivation_key_agreement#c.PSA_KEY_DERIVATION_INPUT_SECRET) as a third option.

mpg replied:
In the future we want to use `input_key` in a lot more places, included here. That's Goal 3 "Allow isolation of short-term secrets (for example, TLS session keys)." in the [strategy document](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/architecture/psa-migration/strategy.md). We're no yet to the point where doing it is in the next step, but we should make sure we're not blocking that path.

mprse replied:
Yes, I have access, so the design under PR https://github.com/ARM-software/psa-crypto-api/pull/549 is valid and I need to adapt implementation to it?

mpg replied:
I think so, unless we're expecting more review on the psa-crypto-api PR - @gilles-peskine-arm ?

gilles-peskine-arm replied:
Any further review would likely be from the Mbed TLS team — probably you two.

So @mprse please follow https://github.com/ARM-software/psa-crypto-api/pull/549, but feel free to request changes there if you think something isn't right. Since you're the first to implement it, you may well discover that what I wrote on paper doesn't work in practice, or isn't precise enough.

mprse commented:
Closing this one as work on this task is continued in PR https://github.com/Mbed-TLS/mbedtls/pull/5726.
